vpr/src/place/manual_move_generator.h

```diff
@@ -0,0 +1,13 @@
+#ifndef VPR_MANUAL_MOVE_GEN_H
+#define VPR_MANUAL_MOVE_GEN_H
+#include "move_generator.h"
+#include "globals.h"
+
+void mmg_get_manual_move_info(ManualMoveInfo mmi);
+
+class ManualMoveGenerator : public MoveGenerator {
```

Comment:
If you can modify the original move generator to achieve your manual move purpose, then this derived class is not needed.

Reply:
Ok, in that case, I think the best solution is to fully utilize the inheritance, and pass in the manual move generator starting from the placement_inner_loop() function. In this way, you can move the branching to the try_place() level, instead of in the try_swap() level.

Reply:
Also, if you are sticking with making a derived class, then I think making the code work is more important and should be done first. It's better to leave the duplicated code alone for now.

Reply:
After giving it more thought, I think the branching in try_swap might be necessary, since you are displaying the graphics and also generating two move outcomes (one from VPR and one from user). But you can still choose to pass in the manual move generator using inheritance.

Basically somewhere in placement_inner_loop's argument list, you have a type ```MoveGenerator& move_generator```
Changing it to a pointer type ```MoveGenerator* move_generator``` will allow you to pass in manual move generator here, having ```MoveGenerator*``` function as a base pointer.